backend/core/src/applications/applications.controller.ts

```diff
@@ -8,64 +8,92 @@ import {
+import AuthzGuard from "../auth/authz.guard"
+import { ResourceType } from "../auth/resource_type.decorator"
+import { AuthzService } from "../auth/authz.service"
 
 @Controller("applications")
 @ApiTags("applications")
 @ApiBearerAuth()
 @UseInterceptors(ClassSerializerInterceptor)
+@ResourceType("application")
+@UseGuards(OptionalAuthGuard, AuthzGuard)
 export class ApplicationsController {
-  constructor(private readonly applicationsService: ApplicationsService) {}
+  constructor(
+    private readonly applicationsService: ApplicationsService,
+    private readonly authzService: AuthzService
+  ) {}
 
   @Get()
-  @UseGuards(DefaultAuthGuard)
   @ApiOperation({ summary: "List applications", operationId: "list" })
   @UseInterceptors(new TransformInterceptor(ApplicationDto))
-  async list(@Query() params: ApplicationsListQueryParams): Promise<ApplicationDto[]> {
-    return await this.applicationsService.list(params)
+  async list(
+    @Request() req,
+    @Query() params: ApplicationsListQueryParams
+  ): Promise<ApplicationDto[]> {
+    // TODO: Do we want to return all applications for admin users?
+    return await this.applicationsService.list(params, req.user)
   }
 
   @Post()
   @ApiOperation({ summary: "Create application", operationId: "create" })
-  @UseGuards(OptionalAuthGuard)
   @UseInterceptors(new TransformInterceptor(ApplicationDto))
-  async create(@Body() applicationCreateDto: ApplicationCreateDto): Promise<ApplicationDto> {
-    return await this.applicationsService.create(applicationCreateDto)
+  async create(
+    @Request() req,
+    @Body() applicationCreateDto: ApplicationCreateDto
+  ): Promise<ApplicationDto> {
+    return await this.applicationsService.create({ ...applicationCreateDto, user: req.user })
   }
 
   @Get(`:applicationId`)
-  @UseGuards(DefaultAuthGuard)
   @ApiOperation({ summary: "Get application by id", operationId: "retrieve" })
   @UseInterceptors(new TransformInterceptor(ApplicationDto))
-  async retrieve(@Param("applicationId") applicationId: string): Promise<ApplicationDto> {
-    return await this.applicationsService.findOne(applicationId)
+  async retrieve(
+    @Request() req,
+    @Param("applicationId") applicationId: string
+  ): Promise<ApplicationDto> {
+    const app = await this.applicationsService.findOne(applicationId)
+    await this.authorizeUserAction(req.user, app, "read")
+    return app
   }
 
   @Put(`:applicationId`)
-  @UseGuards(DefaultAuthGuard)
   @ApiOperation({ summary: "Update application by id", operationId: "update" })
   @UseInterceptors(new TransformInterceptor(ApplicationDto))
   async update(
+    @Request() req,
     @Param("applicationId") applicationId: string,
     @Body() applicationUpdateDto: ApplicationUpdateDto
   ): Promise<ApplicationDto> {
-    return await this.applicationsService.update(applicationUpdateDto)
+    const app = await this.applicationsService.findOne(applicationId)
+    await this.authorizeUserAction(req.user, app, "edit")
```

Comment:
What do you think about make those "edit", "read", "delete", "create" strings an Enum?